### PR TITLE
Add ability to wait for all instances in an application to be RUNNING

### PR DESCRIPTION
**Pull request by cpettitt-confluent**
Example output on failure:

```
java.lang.AssertionError: Application did not reach a RUNNING state for all streams instances. Non-running instances: {org.apache.kafka.streams.KafkaStreams@77258e59=REBALANCING, org.apache.kafka.streams.KafkaStreams@22295ec4=REBALANCING}

	at org.junit.Assert.fail(Assert.java:89)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.startApplicationAndWaitUntilRunning(IntegrationTestUtils.java:789)
	at org.apache.kafka.streams.integration.OptimizedKTableIntegrationTest.standbyShouldNotPerformRestoreAtStartup(OptimizedKTableIntegrationTest.java:112)
```

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (cpettitt-confluent) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:801]: Note that the streams toString is just the object and hash (e.g. `org.apache.kafka.streams.KafkaStreams@77258e59`). It would be nice if there were a better way to identify the instance, but I couldn't find anything in the public API.

File: streams/src/test/java/org/apache/kafka/streams/integration/OptimizedKTableIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/OptimizedKTableIntegrationTest.java]: Thinking about it, should we switch to use `Duration` (would apply to my own PR, too).

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: nit: add missing `final` (I assume checkstyle will fail)

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:805]: Just to clarify: when we call await() here, we release `stateLock`, right?

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:801]: Maybe we could use `client.id` (if set?) from `StreamsConfig`?

**Review by mjsax — COMMENTED** [review]
Overall LGTM

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (cpettitt-confluent) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:805]: That's exactly how it works. Some more details here, if you're curious: https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/Condition.html#awaitNanos(long).

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (cpettitt-confluent) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: Surprisingly it didn't and the Jenkins tests didn't fail on it either. Maybe there is a bug in how it handles compound exception expressions. I'll fix it regardless.

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (cpettitt-confluent) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:801]: That would be great - client.id is what I was going for - but I don't see a good way to get the config out of the KafkaStreams either (besides reflection :)). Did I miss anything?

File: streams/src/test/java/org/apache/kafka/streams/integration/OptimizedKTableIntegrationTest.java
Comment (cpettitt-confluent) [reply streams/src/test/java/org/apache/kafka/streams/integration/OptimizedKTableIntegrationTest.java]: Will do.

**Comment by cpettitt-confluent** [comment]
New patch up that addresses above requests (use Duration, use final).

**Comment by cpettitt-confluent** [comment]
11/2.12 and 11/2.13 passed.

8/2.11 failed with 1 known flaky test (https://issues.apache.org/jira/browse/KAFKA-8967): testCreateTopicsResponseMetadataAndConfig

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:751]: We recently merge a fix for EosIntegrationTest that also starts two instances -- can we update this test to use this new method in this PR? https://github.com/apache/kafka/blob/trunk/streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java#L467-L468

Also, the other PR introduced `StreamsTestUtils#startKafkaStreamsAndWaitForRunningState(...)` that takes a long as timeout -- to align with this PR, we could change it to `Duration` as well.

We can also do a follow up PR if you don't want to block this PR by both.

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:801]: Well, maybe we can update `KafkaStreams#toString()` to access the `client.id` ?

**Comment by mjsax** [comment]
Java 11 / 2.13 and 2.12 passed.

Java8: `kafka.api.SaslSslAdminClientIntegrationTest.testCreateTopicsResponseMetadataAndConfig`

**Review by mjsax — APPROVED** [review]
Overall LGTM.

Call for second review @ableegoldman

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: Maybe it was suppressed? Or could checkstyle actually be missing things, I remember it failed to pick up an unused import a while ago 🤷‍♀

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:805]: @mjsax haven't you had to do the `Job Scheduler` interview question? :P

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/CompositeStateListener.java
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/CompositeStateListener.java:30]: nit: execution of **all** listeners?

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/CompositeStateListener.java
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/CompositeStateListener.java]: nit: **thrown** exception?

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:814]: I just have to say, personally I think reflection seems like overkill here, but I take it we're just trying to avoid the need to go through a KIP?
It seems like a potentially useful API, although admittedly I've never seen it requested specifically. Just wondering what the thoughts/status is here: are we considering adding it as an official API at some later point?

**Review by ableegoldman — APPROVED** [review]
Thanks, this'll be a useful tool in the never-ending fight against flaky tests -- left a few very minor comments but pretty much LGTM

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:814]: I would be open to add this to the public API -- but short term, reflections are the only "hack" we can use.

**Comment by cpettitt-confluent** [comment]
> Java 11 / 2.13 and 2.12 passed.
> 
> Java8: `kafka.api.SaslSslAdminClientIntegrationTest.testCreateTopicsResponseMetadataAndConfig`

The test failure (`kafka.api.SaslSslAdminClientIntegrationTest.testCreateTopicsResponseMetadataAndConfig`) is a known flaky test: https://issues.apache.org/jira/browse/KAFKA-8967.

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/CompositeStateListener.java
Comment (cpettitt-confluent) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/CompositeStateListener.java:30]: There's a fair amount of precedence for this phrase in the Kafka Streams source and it reads a little clearer to me. You're absolutely correct from a logical perspective.

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/CompositeStateListener.java
Comment (cpettitt-confluent) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/CompositeStateListener.java]: Thanks! I'll post a new PR with this fixed.

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:751]: Do we need both startKafkaStreamsAndWaitForRunningState and this? Can we just consolidate them into a single function?

**Review by guozhangwang — COMMENTED** [review]
Made a pass on the PR, thanks for the great improvement @cpettitt-confluent !!

**Comment by cpettitt-confluent** [comment]
Assuming we're looking for green tests to merge this? The run pre-one-character-doc-change failed on a single flaky test. retest this please

**Review by guozhangwang — APPROVED** [review]
LGTM. Leaving to @mjsax to merge.

**Comment by mjsax** [comment]
Java 11 / 2.13 passed.
Java 11 / 2.12 and Java 8 failed. Results are not available any longer.

Retest this please.

**Comment by cpettitt-confluent** [comment]
@mjsax It might be worth pointing out that this is a test-only change and the code is only used by one test, which has not been failing. So the test failures must therefore be flaky and unrelated.

**Comment by mjsax** [comment]
Java 8: `kafka.log.LogCleanerIntegrationTest.testIsThreadFailed`
Java 11 / 2.13:
```
kafka.api.DelegationTokenEndToEndAuthorizationTest.testNoDescribeProduceOrConsumeWithoutTopicDescribeAcl
kafka.log.LogCleanerIntegrationTest.testIsThreadFailed
org.apache.kafka.streams.integration.SmokeTestDriverIntegrationTest.shouldWorkWithRebalance
```
Java 11 / 2.12:
```
kafka.api.SaslScramSslEndToEndAuthorizationTest.testNoDescribeProduceOrConsumeWithoutTopicDescribeAcl
kafka.api.SaslSslAdminClientIntegrationTest.testCreateTopicsResponseMetadataAndConfig
```

Different set of tests failed. Merging.

**Comment by mjsax** [comment]
Merged to `trunk` and cherry-picked to `2.4`

**Pull request merged**
